fetch.ts

```diff
@@ -28,18 +35,50 @@ const FETCH_OPTIONS: FetchOptions = {
 export function getRetryingFetch(retryOnFail: boolean, maxRetries: number) {
   return async function (url: string, options = FETCH_OPTIONS) {
     let retries = 0;
-    let response: Response | undefined;
+    let retryDelay = 2000; // 2 seconds
+    const maxDelay = 30000; // 30 seconds
+    const timeout = 30000; // 30 seconds
+    let response: ExtendedResponse | undefined | null;
     let error: unknown;
     do {
+      if (retries > 0) {
+        console.log(`Retrying (${retries}) after ${retryDelay / 1000} seconds`);
+        await new Promise((resolve) => setTimeout(resolve, retryDelay));
+        retryDelay = Math.min(retryDelay * 2, maxDelay);
+        console.log("Retrying...");
+      }
       try {
-        response = await fetch(url, options);
+        const controller = new AbortController();
+        const timeoutId = setTimeout(() => controller.abort(), timeout);
+        if (MANUAL_REDIRECTION.includes(url)) {
+          response = await fetch(url, { ...options, redirect: "manual", signal: controller.signal });
+          response.isRedirected = true;
+        } else {
+          response = await fetch(url, { ...options, signal: controller.signal });
+          response.isRedirected = response.redirected;
+        }
+        clearTimeout(timeoutId);
       } catch (err) {
         error = err;
         if (!retryOnFail) break;
-        console.log(`Retrying (${retries + 1})`);
+
+        // Check if the error is due to timeout
+        if (err.name === "AbortError") {
+          console.error(`Timeout (${timeout / 1000} seconds) reached`);
+        }
+      }
+
+      if (
+        retries <= maxRetries &&
+        response != null &&
+        // 408 Request Timeout
+        // 5xx Internal Server Error
+        (response.status === 408 || response.status > 500)
+      ) {
+        response = null; // Retry!
       }
       retries++;
-    } while (retries < maxRetries && response == null);
+    } while (retries <= maxRetries && response == null);
```

Comment:
```suggestion
    } while (retries < maxRetries && response == null);
```

~~`<=` makes 6 retries in total, ignoring `maxRetries` (= 5), because `retries` starts from `0`~~. My mistake.